in noImplicitReturns mode, also disallow "return;"

martine opened the pull request:
In --noImplicitReturns mode, if a function specifies a return type,
disallow empty "return;" statements.

Fixes #5916.

---

This is my first TypeScript change, so please review with caution.

Some things wasn't sure about, checked off the resolved ones:
- [x] Should I add more tests for other cases?  E.g. a function that is tagged with `void` return etc.
- [ ] Should I use a different error message?
- [x] Maybe my check against Any here is wrong?  I copied it from elsewhere.

msftclas commented:
Hi __@martine__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

martine commented:
The Travis build appears to have failed due to a lint error (?).

I don't see any mention of lint in `CONTRIBUTING.md`, suggestions welcome!

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13880,6 +13880,11 @@ namespace ts {
                             checkTypeAssignableTo(exprType, returnType, node.expression);
                         }
                     }
+                } else {
```
Else on the next line (we link to a style guide in CONTRIBUTING.md)

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13880,6 +13880,11 @@ namespace ts {
+                } else {
+                  if (compilerOptions.noImplicitReturns && !maybeTypeOfKind(returnType, TypeFlags.Any | TypeFlags.Void)) {
+                    // The function has a return type, but the return statement doesn't have an expression.
+                    error(node, Diagnostics.Not_all_code_paths_return_a_value);
```
4 spaces per indentation

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -13880,6 +13880,11 @@ namespace ts {
                         }
                     }
+                } else {
+                  if (compilerOptions.noImplicitReturns && !maybeTypeOfKind(returnType, TypeFlags.Any | TypeFlags.Void)) {
```
This can be an `else if`

DanielRosenwasser commented:
Can you rename the test file? It's not that there's an empty statement, it's that the return statement lacks a return expression.

Other than that and a few style nits, this looks good. @vladima, can you also take a look?

DanielRosenwasser commented:
Also, `jake lint` will run our linter.

martine commented:
So embarrassing that I got tabs wrong, it's such a pet peeve of mine when others get it wrong!

I think I have fixed the issues you brought up.  I'm still not quite sure I'm checking for the right thing (e.g. isTypeOfKind vs maybeTypeOfKind) but I added some more test cases to verify it makes some sense.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -13881,6 +13881,10 @@ namespace ts {
                         }
                     }
                 }
+                else if (compilerOptions.noImplicitReturns && !isTypeOfKind(returnType, TypeFlags.Void)) {
```
void | any

martine replied:
With your suggestion, this code is now accepted.  Is that expected?  (I don't have a good intuition about how "any" ought to behave, so I just wanted to double check.)
```    
    function isMissingReturnExpression2(): any {
      return;
    }
```

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -13881,6 +13881,10 @@ namespace ts {
                         }
                     }
                 }
+                else if (compilerOptions.noImplicitReturns && !isTypeOfKind(returnType, TypeFlags.Void | TypeFlags.Any)) {
```
please use maybeTypeOfKind to support something like `T | void`.

mhegazy replied:
also looks like we do not handel this correctelly in 10804:
```ts
if (inferredReturnType === voidType || isTypeAny(inferredReturnType)) {
```

should also use `maybeTypeOfKind` as well.

mhegazy replied:
a test would be appreciated as well.

martine replied:
Aw, I had a test for that as well.  OK, one more rebaseline...

mhegazy commented:
also can you add a test for an implicit return type that is inferred from the function body:

```ts
function f(x) {
   if (x) {
       return 0;
   }
   else {
       return;
   }
}
```

martine commented:
Done.  I also fixed the tabs in my test code (adding the "if" statement made me realize I had it wrong).

vladima commented:
:+1:

mhegazy commented:
Wow.. @traviceCI is taking its time for this commit..

DanielRosenwasser commented:
The suspense...

mhegazy commented:
thanks @martine!
